server/src/services/vault-manager.js

```diff
@@ -0,0 +1,247 @@
+}
+
+// encrypts full contents of path or throws if encrypted subfolder exists
+exports.encryptFolder = (path, key) => {
+  path = exports.getVaultPath(path)
+
+  // check if subfolder already encrypted
+  if (walk.walkSync(path).find(f => f.name === '.locked')) {
+    throw new HTTPError(400, 'Path already encrypted')
+  }
+
+  // encrypt all files not starting with '.'
+  const files = walk.walkSync(path, WALK_SETTINGS)
+  for (const file of files) {
+    const data = fs.readFileSync(file.path)
+    exports.encryptData(data, key, file.path)
+  }
+
+  // add lock file for key verification
+  const lock = ospath.join(path, '.locked')
+  const data = aes.encrypt(LOCK_TEXT, key)
+  fs.writeFileSync(lock, data)
+}
+
+// decrypts data with key and writes result to outPath, or returns if no outPath
+exports.decryptData = (data, key, outPath) => {
+  const dec = aes.decrypt(data, key)
+  if (!outPath) return dec
+  fs.writeFileSync(outPath, dec)
+}
+
+// decrypts full contents of path or throws false if key invalid
+exports.decryptFolder = (path, key) => {
+  path = exports.getVaultPath(path)
+
+  if (!exports.isPathLocked(path)) {
+    throw new HTTPError(400, 'Path is not locked')
+  }
+  if (!exports.isKeyValid(path, key)) {
+    throw new HTTPError(400, 'Key is not valid')
+  }
+
+  // decrypt all files not starting with '.'
+  const files = walk.walkSync(path, WALK_SETTINGS)
+  for (const file of files) {
+    const data = fs.readFileSync(file.path)
+    exports.decryptData(data, key, file.path)
+  }
+
+  // remove lock file
+  const lock = ospath.join(path, '.locked')
+  fs.removeSync(lock)
+}
+
+// deletes a path
+exports.deletePath = path => {
+  if (!path || path === 'shared') {
+    throw HTTPError.FORBIDDEN
+  }
+
+  path = exports.getVaultPath(path)
+  fs.removeSync(path)
+}
+
+// returns file data of path, decrypted with key if exists
+exports.getFile = (path, key) => {
+  path = exports.getVaultPath(path)
+  let data = fs.readFileSync(path)
+
+  if (key !== undefined) {
+    data = exports.decryptData(data, key)
+  }
+
+  return data
+}
+
+// return encryption root of path, or null if not encrypted
+exports.getLockedPath = path => {
+  if (path.startsWith(FILES_DIR)) {
+    path = path.slice(FILES_DIR.length)
+  }
+
+  const parts = path.split(ospath.sep)
+  let subPath = ''
+  for (const part of parts) {
+    subPath = ospath.join(subPath, part)
+    path = ospath.join(FILES_DIR, subPath)
+
+    if (fs.existsSync(ospath.join(path, '.locked'))) {
+      return subPath + ospath.sep
+    }
+  }
+
+  return null
+}
+
+// return full path of item in vault
+exports.getVaultPath = (subPath, enforceExists) => {
+  const path = ospath.join(FILES_DIR, subPath || '')
+  if (!exports.isSafePath(path, undefined, enforceExists)) {
+    throw HTTPError.NOT_FOUND
+  }
+
+  return path
+}
+
+// returns true if key is correct to decrypt
+exports.isKeyValid = (path, key) => {
+  path = exports.getLockedPath(path)
+  if (path === null) return true
```

Comment:
why true?

Reply:
if path is not locked, return `true` because no key is needed. `false` is used for auth errors

Reply:
null path always means not locked?

Reply:
just read it, jk